app/assets/src/components/views/SampleUploadFlow/UploadProgressModal.jsx

```diff
@@ -0,0 +1,486 @@
+        </React.Fragment>
+      );
+    }
+
+    // While local samples are being uploaded.
+    if (this.someLocalSamplesInProgress()) {
+      return (
+        <React.Fragment>
+          <div className={cs.title}>
+            Uploading {samples.length} sample{samples.length !== 1 && "s"} to{" "}
+            {project.name}
+          </div>
+          <div className={cs.subtitle}>
+            Please stay on this page until your upload completes.
+          </div>
+        </React.Fragment>
+      );
+    }
+
+    // If any local samples failed.
+    if (this.someLocalSamplesFailed()) {
+      return this.renderFailedSamplesTitle();
+    }
+
+    // If all local samples succeeded.
+    return (
+      <React.Fragment>
+        <div className={cs.titleWithIcon}>
+          <CheckmarkIcon className={cs.checkmarkIcon} />
+          All samples uploaded successfully!
+        </div>
+      </React.Fragment>
+    );
+  };
+
+  renderSampleLoadingBar = sample => {
+    const uploadPercentage = this.getSampleUploadPercentage(sample);
+
+    return (
+      <div className={cs.loadingBarBackground}>
+        <div
+          className={cs.loadingBar}
+          style={{ width: `${uploadPercentage * 100}%` }}
+        />
+      </div>
+    );
+  };
+
+  renderViewProjectButton = () => {
+    const { project } = this.props;
+    const { uploadComplete } = this.state;
+    if (!uploadComplete) {
+      return null;
+    }
+
+    return (
+      <a className={cs.link} href={`/home?project_id=${project.id}`}>
+        <PrimaryButton
+          text="Go to Project"
+          rounded={false}
+          onClick={() =>
+            logAnalyticsEvent("UploadProgressModal_to-project-button_clicked", {
+              projectId: project.id,
+              projectName: project.name,
+            })
+          }
+        />
+      </a>
+    );
+  };
+
+  render() {
+    const { samples, uploadType } = this.props;
+    const { failedSampleNames } = this.state;
+
+    return (
+      <Modal open tall narrow className={cs.uploadProgressModal}>
+        <div className={cs.header}>
+          <UploadIcon className={cs.uploadIcon} />
+          {this.renderTitle()}
+        </div>
+        {uploadType === "local" && (
+          <div className={cs.sampleList}>
+            {samples.map(sample => (
+              <div key={sample.name} className={cs.sample}>
+                <div className={cs.sampleHeader}>
+                  <div className={cs.sampleName}>{sample.name}</div>
+                  <div className={cs.fill} />
+                  <div className={cs.sampleStatus}>
+                    {this.renderSampleStatus(sample)}
+                  </div>
+                </div>
+                {this.renderSampleLoadingBar(sample)}
+              </div>
+            ))}
+          </div>
+        )}
+        {failedSampleNames.length > 0 && (
+          <div className={cs.failedSamples}>
+            Failed samples: {take(3, failedSampleNames).join(", ")}
```

Comment:
I'm going to do a constant because it feels a little too specific to expose as a prop without a clear use case. `numFailedSamplesToDisplay`. It's easy to refactor later if something needs it.